src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Extensions/ApplicationInsightsServiceCollectionExtensions.cs

```diff
@@ -74,27 +74,29 @@ public static IServiceCollection AddApplicationInsights(this IServiceCollection
                 var includedActivities = dependencyCollector.IncludeDiagnosticSourceActivities;
                 includedActivities.Add("Microsoft.Azure.ServiceBus");
 
-                dependencyCollector.EnableW3CHeadersInjection = options.EnableW3CDistributedTracing;
+                dependencyCollector.EnableW3CHeadersInjection = options.HttpAutoCollectionOptions.EnableW3CDistributedTracing;
                 return dependencyCollector;
             });
 
-            services.AddSingleton<ITelemetryModule, RequestTrackingTelemetryModule>(provider =>
+            services.AddSingleton<ITelemetryModule>(provider =>
             {
                 var options = provider.GetService<IOptions<ApplicationInsightsLoggerOptions>>().Value;
-                var appIdProvider = provider.GetService<IApplicationIdProvider>();
+                if (options.HttpAutoCollectionOptions.CollectExtendedHttpTriggerInformation)
+                {
+                    var appIdProvider = provider.GetService<IApplicationIdProvider>();
 
-                var requestCollector =
-                    new RequestTrackingTelemetryModule(appIdProvider)
+                    return new RequestTrackingTelemetryModule(appIdProvider)
                     {
                         CollectionOptions = new RequestCollectionOptions
                         {
                             TrackExceptions = false, // webjobs/functions track exceptions themselves
-                            EnableW3CDistributedTracing = options.EnableW3CDistributedTracing,
-                            InjectResponseHeaders = options.EnableResponseHeaderInjection
+                            EnableW3CDistributedTracing = options.HttpAutoCollectionOptions.EnableW3CDistributedTracing,
+                            InjectResponseHeaders = options.HttpAutoCollectionOptions.EnableResponseHeaderInjection
                         }
                     };
+                }
 
-                return requestCollector;
+                return new NoopTelemetryModule<RequestTrackingTelemetryModule>();
```

Comment:
@brettsam, do you know a better way for it to avoid adding anything when `CollectExtendedHttpTriggerInformation` is false?

Reply:
You can return null here... but whenever you call GetServices<>(), you have to be prepared for one of the services in the collection to be null. But that feels a little messy -- I kinda wish if you returned null it'd just ignore that registered service completely.

We take this approach elsewhere (having some no-op/null service that we return instead). @fabiocav -- is there a "right" way to do this in a factory where you decide you don't want that service injected due to options? I went hunting through some of the .NET Core repos but couldn't find a good example of this pattern.

Reply:
This would be my recommended approach, to be honest. Otherwise (as @brettsam) pointed out, you put the burden of checking on the consumers and it makes constructor injection problematic (may not be the case with this if you're taking a collection)